book/src/plumbing/jars_and_ingredients.md

```diff
@@ -167,15 +159,16 @@ We can then do things like ask, "did this input change since revision R?" by
 * using the ingredient index to find the route and get a `&dyn Ingredient<DB>`
 * and then invoking the `maybe_changed_since` method on that trait object.
 
-### HasJarsDyn
+### `HasJarsDyn`
 
 There is one catch in the above setup.
-We need the database to be dyn-safe, and we also need to be able to define the database trait and so forth without knowing the final database type to enable separate compilation.
+We need the database to be dyn-safe [what does this mean?], and we also need to be able to define the database trait without knowing the final database type to enable separate compilation.
```

Comment:
It's not entirely clear to me what "dyn-safe" means here. An example or a definition would help.

Reply:
Ah, yeah, that's "Rust jargon". It means "usable with dyn". I'm trying to find something to link to and failing! That...seems like a gap in the Rust reference.

Reply:
Notably, a trait with generic methods cannot be used `dyn`. Let me see if I can rewrite this.

```suggestion
The user's code always interacts with a `dyn crate::Db` value, where `crate::Db` is the trait defined by the jar; the `crate::Db` trait extends `salsa::HasJar` which in turn extends `salsa::Database`.
Ideally, we would have `salsa::Database` extend `salsa::HasJars`, which is the main trait that gives access to the jars data.
But we don't want to do that because `HasJars` defines an associated type `Jars`, and that would mean that every reference to `dyn crate::Db` would have to specify the jars type using something like `dyn crate::Db<Jars = J>`.
This would be unergonomic, but what's worse, it would actually be impossible: the final Jars type combines the jars from multiple crates, and so it is not known to any individual jar crate.
To workaround this, `salsa::Database` in fact extends *another* trait, `HasJarsDyn`, that doesn't reveal the `Jars` or ingredient types directly, but just has various method that can be performed on an ingredient, given its `IngredientIndex`.
```

Reply:
More verbose, but hopefully clearer!

Reply:
> It's not entirely clear to me what "dyn-safe" means here. An example or a definition would help.

> Ah, yeah, that's "Rust jargon". It means "usable with dyn". I'm trying to find something to link to and failing! That...seems like a gap in the Rust reference.

Isn't it https://doc.rust-lang.org/reference/items/traits.html#object-safety ?